Dashboard TOC view

MVarshini opened the pull request:
PBENCH-1217

Fix the dashboard TOC view

dbutenhof reviewed (changes requested):
Here's one thing that didn't work in the original code, and which still doesn't work:

1. On the browsing page, click on the first dataset, `fio_rw_2018.02.01T22.40.57`
2. Click on the first directory, `1-rw-4KiB`
3. Click on the first directory, `sample1`

You'll see an error pop up with `DatasetsContents: Postprocessing error returning 404: "No directory '/1-rw-4KiBsample1' in '22a4bc5748b920c6ce271eb68f08d91c' contents."`

The problem is that the correct URL is `/api/v1/datasets/22a4bc5748b920c6ce271eb68f08d91c/contents/1-rw-4KiB/sample1`, but the dashboard hasn't added the "/" between the two directories.

(And while this doesn't really come under the heading of "fixing", it's always bothered me a bit that the files in a directory are duplicated in the left and right columns of the display. My original vision for the TOC was an indented cascade with disclosure triangles to open nested directory lists, so that we'd build an indented list of directories and files they contain in one column, possibly with a right tab to display the *contents* of files. But I'm not asking for a full page redesign in this PR, just making it work as previously designed. 😆 )

webbnh reviewed (changes requested):
This looks like a nice clean-up, however, you missed a reference to `isLoading` which you should fix.

Beyond that, if you're so inclined, I have a couple of nits/suggestions for you to consider.

MVarshini commented on dashboard/src/reducers/tableOfContentReducer.js:
```diff
@@ -27,7 +26,6 @@ const TableOfContentReducer = (state = initialState, action = {}) => {
         searchSpace: payload.files,
         tableData: payload.files,
         contentData: payload,
-        isLoading: false,
```
Removed the `isLoading` flag

MVarshini commented on dashboard/src/modules/components/TableOfContent/index.jsx:
```diff
@@ -179,6 +177,7 @@ const TableOfContent = () => {
     setTableData(stack[0].files);
+    dispatch(updateContentData(stack[0]));
     setSearchSpace(stack[0].files);
     setBreadCrumb(initialBreadcrumb([]));
```
Gotcha!

MVarshini commented:
> Here's one thing that didn't work in the original code, and which still doesn't work:
> 
> 1. On the browsing page, click on the first dataset, `fio_rw_2018.02.01T22.40.57`
> 2. Click on the first directory, `1-rw-4KiB`
> 3. Click on the first directory, `sample1`
> 
> You'll see an error pop up with `DatasetsContents: Postprocessing error returning 404: "No directory '/1-rw-4KiBsample1' in '22a4bc5748b920c6ce271eb68f08d91c' contents."`
> 
> The problem is that the correct URL is `/api/v1/datasets/22a4bc5748b920c6ce271eb68f08d91c/contents/1-rw-4KiB/sample1`, but the dashboard hasn't added the "/" between the two directories.
> 
> (And while this doesn't really come under the heading of "fixing", it's always bothered me a bit that the files in a directory are duplicated in the left and right columns of the display. My original vision for the TOC was an indented cascade with disclosure triangles to open nested directory lists, so that we'd build an indented list of directories and files they contain in one column, possibly with a right tab to display the _contents_ of files. But I'm not asking for a full page redesign in this PR, just making it work as previously designed. laughing )

Updated the URL. Also, added download option. As metioned, will redesign the page later

dbutenhof commented on dashboard/src/modules/components/TableOfContent/index.jsx:
```diff
@@ -207,171 +209,173 @@ const TableOfContent = () => {
+                      <Td dataLabel={file.type}>{file.type}</Td>
+                      <Td dataLabel={file.uri}>
+                        <a
+                          className="download-icon"
+                          href={file.uri}
+                          target="_blank"
+                          rel="noreferrer"
```
This is cool, and I think reveals a bug in the server 😆 

I'm getting weird downloaded file names like `fio_rw_2018.02.01T22.40.57_metadata.log`, which I suspect comes from the fact that our `Content-Disposition` header is specifying the file *path* `fio_rw_2018.02.01T22.40.57/metadata.log` instead of just the name.

dbutenhof commented on dashboard/src/modules/components/TableOfContent/index.jsx:
```diff
@@ -176,29 +178,29 @@ const TableOfContent = () => {
   };
   const attachBreadCrumbs = (data, firstHierarchyLevel) => {
-    breadCrumbLabels.push(data);
+    breadCrumbLabels.push(data.name);
```
This doesn't seem to be working; or at least not as I'd expect.

We start out with "Root >", and clicking on "1-rw-4GiB" changes the breadcrumb to "Root > 1-rw-4GiB" as I'd expect ... but clicking on "sample1" doesn't change the breadcrumb display again, nor does clicking on "csv" inside that: it always stays "Root > 1-rw-4GiB" until I click on "Root" to reset.

Also, I'd expected that each breadcrumb would be an active link, but only "Root" is a link. For example, I'd expect to see "Root > 1-rw-4GiB > sample1 > csv", and that I could click on "Root", "1-rw-4GiB", or "sample1" to get back to that directory.

dbutenhof replied:
Fair enough. I created [PBENCH-1242](https://issues.redhat.com/browse/PBENCH-1242) to track it separately.

dbutenhof reviewed (changes requested):
This is a massive improvement, and I really like everything except the behavior of the breadcrumb list and the download file name. But the first isn't necessarily a deal breaker if it's hard to fix, and it looks like the second is actually an undesirable server behavior that's not your problem. 😆 

Take a look at the breadcrumb logic, please; hopefully we can make this look better so we can navigate back up a nested directory tree rather than only seeing "Root" and the outermost directory!

webbnh reviewed (approved):
OK, there's no longer anything here that I would block the merge for (once Dave says it's working satisfactorily 😉), however I do have some items for you to consider.

The biggest one is a suspicious modification of the Redux store -- I don't know how bad that is, but I suggested a slightly different approach.  Next is another questionable use of the `?:` operator.  Beyond those, I have a couple of coding suggestions.

dbutenhof commented on dashboard/src/modules/components/TableOfContent/index.jsx:
```diff
@@ -70,6 +70,9 @@ const TableOfContent = () => {
   const setTableData = (data) => {
     dispatch(updateTableData(data));
   };
+  const setContnetData = (data) => {
```
Was this supposed to be `setContentData`??

webbnh replied:
I'd never thought of IDEs as _enabling_ misspellings before....  🤯

webbnh reviewed (approved):
Looks OK to me, although you probably _should_ fix [the misspelling](https://github.com/distributed-system-analysis/pbench/pull/3511#discussion_r1287369639).
